map.go

```diff
@@ -1575,6 +1600,26 @@ func (mi *MapIterator) Next(keyOut, valueOut interface{}) bool {
 	return false
 }
 
+// Next decodes the next key and value.
+//
+// Iterating a hash map from which keys are being deleted is not
+// safe. You may see the same key multiple times. Iteration may
+// also abort with an error, see IsIterationAborted.
+//
+// Iterating a Queue/Stack will lookup only the value, as those
+// maps do not use a key.
+//
+// Returns false if there are no more entries. You must check
+// the result of Err afterwards.
+//
+// See Map.Get for further caveats around valueOut.
+func (mi *MapIterator) Next(keyOut, valueOut interface{}) bool {
+	if mi.target.typ.isQueueStack() {
+		return mi.nextQueueMap(valueOut)
+	}
+	return mi.next(keyOut, valueOut)
```

Comment:
I think that we should implement this slightly differently. Now that we have multiple iterator implementations (and I expect more to come), it makes sense to start using an interface with multiple implementations instead.

I suggest renaming `MapIterator`, and making it unexported. Then add a `MapIterator` interface with the `Err() error` and `Next(keyOut, valueOut interface{}) bool` methods. And add your logic as an new stack/queue iterator struct. 

Then make `Map.Iterate` pick the correct iterator.

Reply:
That is perfectly fine by me.  I'm going to create two MapIterator implementations, one for key-value maps `indexedMapIterator`  and one for value-only maps `bufferedMapIterator` . Correct me if the naming is horrible (I also thought about `keylessMapIterator` and `keyedMapIterator`)

Concerning the PerCpu maps, I'd leave as it is, so that the old `MapIterator` (hence the newly `indexedMapIterator`) handles this case.